clash-lib/src/Clash/Primitives/Scaffold.hs

```diff
@@ -0,0 +1,518 @@
+  wires <- mapM (ident . T.pack . name) o
+  inst  <- ident (T.pack $ primitiveName <> "_inst")
+  blk   <- ident (T.pack $ primitiveName <> "_blk")
+
+  getMon $ blockDecl blk $ concat
+    [ [ NetDecl Nothing o' t | (o', t) <- zip wires wiresTy ]
+    , [ InstDecl
+        Comp
+        Nothing
+        (T.pack primitiveName)
+        inst
+        (toClashParameter <$> parameters)
+        (  zipWith instantiate i (fmap (view _1) args)
+        <> zipWith instantiate o (flip Identifier Nothing <$> wires)
+        )
+      , result wires (bbResult bbCtx)
+      ]
+    ]
+ where
+  args    = drop (length domains) (bbInputs bbCtx)
+  ident   = mkUniqueIdentifier Basic
+  wiresTy = fmap toClashType o
+
+  result ws (Identifier r Nothing, resTy@Product{})
+    = Assignment r
+    (DataCon resTy (DC (resTy, 0)) [ Identifier w Nothing | w <- ws ])
+  result ws (Identifier r Nothing, _) | [wire] <- ws
+    = Assignment r (Identifier wire Nothing)
+  result _ t = error $ "scaffoldTemplate: unexpected result type"
+                     ++ show t
+
+-- |
+scaffoldTF
+  :: [Int]
+  -> String
+  -> [Parameter]
+  -> [Name]
+  -> [ScaffoldPort]
+  -> [ScaffoldPort]
+  -> TemplateFunction
+scaffoldTF used primitiveName parameters domains i o = TemplateFunction
+  used
+  (const True)
+  (scaffoldTemplate primitiveName parameters domains i o)
+
+-- | Annotations to associate our 'TemplateFunction's with our Haskell
+-- functions.
+scaffoldAnnotation :: Name -> Name -> HDL -> Q Exp
+scaffoldAnnotation n ntf hdl =
+  [|InlinePrimitive hdl j|]
+ where
+  j = [I.i| [{ "BlackBox" :
+              { "name" : "#{n}"
+              , "kind": "Declaration"
+              , "format": "Haskell"
+              , "templateFunction" : "#{ntf}"
+              }
+          }]
+      |]
+
+pureDefault :: Exp
+pureDefault = VarE 'pure `AppE` VarE 'def
+
+knownDomains :: [Name] -> [Type]
+knownDomains = fmap (AppT (ConT ''C.KnownDomain) . VarT)
+
+applyDomains :: [Name] -> Type -> Type
+applyDomains = flip (foldl AppT) . fmap VarT
+
+-- | Builds the Haskell datatypes "datatypeNameI" and "datatypeNameO",
+-- both of which are parametric on the domains of the ports.
+makeDatatypes
+  :: ScaffoldDesc
+  -> ([TyVarBndr],[Name])
+  -> ([ScaffoldPort], [Name])
+  -> ([ScaffoldPort], [Name])
+  -> [Dec]
+makeDatatypes desc (kinds,domains) (i,ni) (o,no) =
+  [ build (datatypeNameI desc) (zipWith mkRec i ni) [''Generic]
+  , build (datatypeNameO desc) (zipWith mkRec o no) [''Generic]
+
+  , SigD (functionNameI desc)
+    $ ForallT kinds (knownDomains domains)
+    $ foldr (AppT . AppT ArrowT) retTy (ty <$> iclks)
+  , FunD (functionNameI desc) [Clause (VarP <$> argNames) (NormalB (
+      foldl AppE
+        (foldl AppE (ConE (datatypeNameI desc)) (VarE <$> argNames))
+        (replicate (length i - length iclks) pureDefault)
+      )) []]
+  ]
+ where
+  b = Bang NoSourceUnpackedness NoSourceStrictness
+  drv = (:[]) . DerivClause Nothing . fmap ConT
```

Comment:
I'm not personally a huge fan of pointfree being used in this way (i.e. `(:[]) .`)